File: course_4/module_2/examples/traveling_salesman.go
package main

import (
	"fmt"
	"math"
)

// TravelingSalesmanProblem represents a solver for the TSP using bitmasking and recursion with memoization.
type TravelingSalesmanProblem struct {
	costMatrix [][]int        // Cost matrix representing the graph.
	numCities  int            // Number of cities (vertices).
	visitedAll int            // Bitmask representing all cities visited.
	memo       map[string]int // Memoization table to store results of subproblems.
}

// NewTravelingSalesmanProblem initializes a new TSP solver with the given cost matrix.
func NewTravelingSalesmanProblem(costMatrix [][]int) *TravelingSalesmanProblem {
	numCities := len(costMatrix)
	visitedAll := (1 << numCities) - 1 // Bitmask where all cities are visited.
	return &TravelingSalesmanProblem{
		costMatrix: costMatrix,
		numCities:  numCities,
		visitedAll: visitedAll,
		memo:       make(map[string]int),
	}
}

// tspRecursive is the core recursive function to solve TSP.
// mask: Bitmask representing visited cities.
// pos: Current position (city).
func (tsp *TravelingSalesmanProblem) tspRecursive(mask, pos int) int {
	// Base case: all cities have been visited.
	if mask == tsp.visitedAll {
		return tsp.costMatrix[pos][0] // Return cost to go back to the starting city.
	}

	// Generate a unique key for the current state (mask and position).
	key := fmt.Sprintf("%d|%d", mask, pos)

	// Check if the result for this state is already memoized.
	if val, exists := tsp.memo[key]; exists {
		return val
	}

	// Initialize the minimum cost for this state.
	ans := math.MaxInt

	// Try visiting all unvisited cities.
	for city := 0; city < tsp.numCities; city++ {
		if mask&(1<<city) == 0 { // Check if city is unvisited.
			// Calculate the new cost.
			newCost := tsp.costMatrix[pos][city] + tsp.tspRecursive(mask|(1<<city), city)
			// Update the minimum cost.
			if newCost < ans {
				ans = newCost
			}
		}
	}

	// Memoize the result for the current state.
	tsp.memo[key] = ans
	return ans
}

// Solve computes the minimum cost of the traveling salesman tour.
func (tsp *TravelingSalesmanProblem) Solve() int {
	// Start from the first city with only the first city visited (mask = 1).
	return tsp.tspRecursive(1, 0)
}

// Main function to demonstrate the usage of the TSP solver.
func main() {
	// Define the cost matrix where costMatrix[i][j] is the cost of traveling from city i to city j.
	costMatrix := [][]int{
		{0, 20, 42, 25},
		{20, 0, 30, 34},
		{42, 30, 0, 10},
		{25, 34, 10, 0},
	}

	// Initialize the TSP solver.
	tsp := NewTravelingSalesmanProblem(costMatrix)

	// Solve the TSP and print the minimum cost.
	minCost := tsp.Solve()
	fmt.Printf("Minimum cost: %d\n", minCost)
}

File: course_4/module_2/examples/traveling_salesman.py
import math

class TravelingSalesmanProblem:
    """
    Solve the Traveling Salesman Problem using bitmasking and recursion with memoization.
    """

    def __init__(self, cost_matrix):
        """
        Initialize the TSP solver.
        :param cost_matrix: A 2D list representing the cost matrix. cost_matrix[i][j] is the cost of traveling from vertex i to vertex j.
        """
        self.cost_matrix = cost_matrix
        self.num_vertices = len(cost_matrix)
        self.VISITED_ALL = (1 << self.num_vertices) - 1
        self.memo = {}

    def tsp(self, mask, pos):
        """
        Recursive function to solve TSP using bitmasking and memoization.
        :param mask: Bitmask representing visited cities.
        :param pos: Current position (city).
        :return: Minimum cost of visiting all cities and returning to the starting point.
        """
        # Base case: all cities have been visited
        if mask == self.VISITED_ALL:
            return self.cost_matrix[pos][0]  # Return to the starting city.

        # Check if result is already memoized
        if (mask, pos) in self.memo:
            return self.memo[(mask, pos)]

        # Try visiting all unvisited cities and take the minimum cost path
        ans = math.inf
        for city in range(self.num_vertices):
            if not (mask & (1 << city)):  # If city is not visited
                new_cost = self.cost_matrix[pos][city] + self.tsp(mask | (1 << city), city)
                ans = min(ans, new_cost)

        # Memoize and return the result
        self.memo[(mask, pos)] = ans
        return ans


# Example usage:
if __name__ == "__main__":
    cost_matrix = [
        [0, 20, 42, 25],
        [20, 0, 30, 34],
        [42, 30, 0, 10],
        [25, 34, 10, 0]
    ]
    tsp_solver = TravelingSalesmanProblem(cost_matrix)
    min_cost = tsp_solver.tsp(1, 0)  # Start from city 0 with only city 0 visited
    print(f"Minimum cost: {min_cost}")
